iceoryx_utils/include/iceoryx_utils/internal/cxx/forward_list.inl

```diff
@@ -0,0 +1,581 @@
+        errorMessage(__PRETTY_FUNCTION__, " iterator of other list can't be used ");
+        std::terminate();
+    }
+
+    const size_type toBeAddedIdx = getHeadOfFreeElementList()->nextIdx;
+
+    if (m_size < CAPACITY && toBeAddedIdx < INVALID_INDEX)
+    {
+        auto newNodePointer = getNodePtrFromIdx(toBeAddedIdx);
+
+        // unlink from freeList
+        getHeadOfFreeElementList()->nextIdx = newNodePointer->nextIdx;
+
+        // data class c'tor
+        new (&(newNodePointer->data)) T(std::forward<ConstructorArgs>(args)...);
+
+        // add to usedList
+        const size_type afterToBeEmplacedIdx = afterToBeEmplacedIter.m_iterListNodeIdx;
+        const size_type emplaceBeforeIdx = getNodePtrFromIdx(afterToBeEmplacedIdx)->nextIdx;
+        newNodePointer->nextIdx = emplaceBeforeIdx;
+        getNodePtrFromIdx(afterToBeEmplacedIdx)->nextIdx = toBeAddedIdx;
+
+        ++m_size;
+
+        return iterator{const_cast<forward_list<T, CAPACITY>*>(afterToBeEmplacedIter.m_cList), toBeAddedIdx};
+    }
+    else
+    {
+        errorMessage(__PRETTY_FUNCTION__, " capacity exhausted ");
+        return end();
+    }
+}
+
+
+template <typename T, int64_t CAPACITY>
+typename forward_list<T, CAPACITY>::iterator
+forward_list<T, CAPACITY>::erase_after(const_iterator beforeToBeErasedIter) noexcept
+{
+    if (this != beforeToBeErasedIter.m_cList)
+    {
+        errorMessage(__PRETTY_FUNCTION__, " iterator of other list can't be used ");
+        std::terminate();
+    }
+
+    size_type beforeToBeErasedIdx = beforeToBeErasedIter.m_iterListNodeIdx;
+    if (!isValidIteratorIndex(beforeToBeErasedIdx))
+    {
+        errorMessage(__PRETTY_FUNCTION__, " iterator is end()");
+        return end();
+    }
+
+    size_type toBeErasedIdx = getNodePtrFromIdx(beforeToBeErasedIdx)->nextIdx;
+
+    if (isValidElementIndex(toBeErasedIdx) && !empty())
+    {
+        auto toErasedNodePointer = getNodePtrFromIdx(toBeErasedIdx);
+        // unlink from usedList
+        getNodePtrFromIdx(beforeToBeErasedIdx)->nextIdx = toErasedNodePointer->nextIdx;
+
+        // data'tor data class
+        toErasedNodePointer->data.~T();
+
+        // add to freeList
+        toErasedNodePointer->nextIdx = getHeadOfFreeElementList()->nextIdx;
+        getHeadOfFreeElementList()->nextIdx = toBeErasedIdx;
+
+        --m_size;
+
+        // Iterator to the element following the erased one, or end() if no such element exists.
+        return iterator{const_cast<forward_list<T, CAPACITY>*>(beforeToBeErasedIter.m_cList),
+                        getNodePtrFromIdx(beforeToBeErasedIdx)->nextIdx};
+    }
+
+    // in case list is empty or iterator reached end()
+    return end();
+}
+
+
+template <typename T, int64_t CAPACITY>
+T& forward_list<T, CAPACITY>::front() noexcept
+{
+    auto iter = begin();
+
+    if (!isValidElementIndex(iter.m_iterListNodeIdx))
+    {
+        errorMessage(__PRETTY_FUNCTION__, " tryed accessing not existing element");
+        std::terminate();
```

Comment:
undefined behaviour in C++ so terminate might be justified (but not necessary since we already checked for it, crash/terminate in undefined behaviour often happens since the check is skipped for efficiency reasons...). This is again the question of how to deal with certain kinds of usage-errors.